src/Client/ClientBase.cpp

```diff
@@ -0,0 +1,1367 @@
+    else
+        std::cout << "Bye." << std::endl;
+}
+
+
+void ClientBase::runNonInteractive()
+{
+    if (!queries_files.empty())
+    {
+        auto process_multi_query_from_file = [&](const String & file)
+        {
+            auto text = getQueryTextPrefix();
+            String queries_from_file;
+
+            ReadBufferFromFile in(file);
+            readStringUntilEOF(queries_from_file, in);
+
+            text += queries_from_file;
+            return processMultiQuery(text);
+        };
+
+        /// Read all queries into `text`.
+        for (const auto & queries_file : queries_files)
+        {
+            for (const auto & interleave_file : interleave_queries_files)
+                if (!process_multi_query_from_file(interleave_file))
+                    return;
+
+            if (!process_multi_query_from_file(queries_file))
+                return;
+        }
+
+        return;
+    }
+
+    String text;
+    if (is_multiquery)
+        text = getQueryTextPrefix();
+
+    if (config().has("query"))
+    {
+        text += config().getRawString("query"); /// Poco configuration should not process substitutions in form of ${...} inside query.
+    }
+    else
+    {
+        /// If 'query' parameter is not set, read a query from stdin.
+        /// The query is read entirely into memory (streaming is disabled).
+        ReadBufferFromFileDescriptor in(STDIN_FILENO);
+        readStringUntilEOF(text, in);
+    }
+
+    if (query_fuzzer_runs)
+        processWithFuzzing(text);
+    else
+        processQueryText(text);
+}
+
+
+static void clearTerminal()
+{
+    /// Clear from cursor until end of screen.
+    /// It is needed if garbage is left in terminal.
+    /// Show cursor. It can be left hidden by invocation of previous programs.
+    /// A test for this feature: perl -e 'print "x"x100000'; echo -ne '\033[0;0H\033[?25l'; clickhouse-client
+    std::cout << "\033[0J"
+                    "\033[?25h";
+}
+
+
+static void showClientVersion()
+{
+    std::cout << DBMS_NAME << " client version " << VERSION_STRING << VERSION_OFFICIAL << "." << std::endl;
+}
+
+
+int ClientBase::main(const std::vector<std::string> & /*args*/)
+{
+    UseSSL use_ssl;
+
+    std::cout << std::fixed << std::setprecision(3);
+    std::cerr << std::fixed << std::setprecision(3);
+
+    if (is_interactive)
+    {
+        clearTerminal();
+        showClientVersion();
+    }
+
+    return mainImpl();
+}
+
+
+void ClientBase::init(int argc, char ** argv)
+{
+    namespace po = boost::program_options;
+
+    /// Don't parse options with Poco library, we prefer neat boost::program_options.
+    stopOptionsProcessing();
+
+    stdin_is_a_tty = isatty(STDIN_FILENO);
+    stdout_is_a_tty = isatty(STDOUT_FILENO);
+    terminal_width = getTerminalWidth();
+
+    Arguments common_arguments{""}; /// 0th argument is ignored.
+    std::vector<Arguments> external_tables_arguments;
+
+    readArguments(argc, argv, common_arguments, external_tables_arguments);
```

Comment:
Why not to leave code for parsing query parameters (and maybe external data) here? `clickhouse-client` will support in and code will be simpler without extra virtual method.